Add readwrite access (a.k.a Readwrite Matryoshka doll)

TeranIvy opened the pull request:
This PR is for issue #25. Summary of changes is below.

1) Added RW access for discontinuous fields and operators. 
2) Added checks to forbid `GH_INC` access for operators and `GH_READWRITE` access for continuous fields (`GH_WRITE` continuous access will be dealt with in #138).
3) Discontinuous RW fields behave like discontinuous W fields (e.g. no halo exchange generated for discontinuous RW -> R dependency) when redundant computation is not involved.
4) With RC the schedule changes because of R in RW discontinuous fields and halo exchanges generated because of it.
5) All modified/added test files are checked for compilation where appropriate (and some related to them which I accidentally started editing).
6) Examples in `dynamo/eg9` are updated for `GH_READWRITE` access.
7) Updated Dynamo0.3 API documentation and made references within more consistent.
8) Updated documentation on stub generation (instead of in #132), testing with compilation and using `pycodestyle` instead of `pep8`.

Notes
- I mainly used existing tests for discontinuous fields where I changed writers' accesses from `GH_WRITE` to `GH_READWRITE` in some of them. The main tests updates are in `dynamo0p3_transformations_test.py` due to different behaviour of RW when RC is implemented (e.g. schedule changes due to added halo exchanges because of R in RW access). I am happy to add/modify the tests.
- I think there are potential duplications in modifications of `def unique_declns_by_intent(self, datatype):` (`psyGen.py`) for RW access (intent behaviour is the same as for `write_args` so perhaps the lists can be joined).
- The modified tests for general CMA/LMA operator writers in `dynamo0p3_cma_test.py` could be done with `pytest.fixture` functionality instead of loops. I can change that if requested.

arporter commented:
@rupertford, do you want to review this one or shall I take it on?

rupertford commented:
I'll do it. You can take a look at the CLAW integration changes :-)

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_asm_field_kernel_mod.F90:
```diff
@@ -1,11 +1,10 @@
-!-------------------------------------------------------------------------------
```
Please add your name

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/argument_mod.F90:
```diff
@@ -37,6 +37,7 @@
 ! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 ! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
+! Modified I. Kavcic Met Office
```
Please add 2018 to the copyright date for changes

TeranIvy replied:
Added.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_any_space_3_mod.f90:
```diff
@@ -30,17 +30,18 @@
 ! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+! Modified I. Kavcic Met Office
```
Please add 2018 to copyright date.

TeranIvy replied:
Added.

TeranIvy commented on examples/dynamo/README:
```diff
@@ -142,7 +142,7 @@ remove and/or change the location and depth of halo exchanges:
     psyclone helmholtz_solver_alg_mod.x90 -s ./redundant_script.py
```
Updated.

TeranIvy commented on examples/dynamo/eg4/enforce_bc_kernel_mod.f90:
```diff
@@ -1,16 +1,14 @@
 ! BSD 3-Clause License
 !
-! Copyright (c) 2017, Science and Technology Facilities Council
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Added.

TeranIvy commented on examples/dynamo/eg4/enforce_bc_kernel_mod.f90:
```diff
@@ -39,7 +37,8 @@
 ! POSSIBILITY OF SUCH DAMAGE.
-!-------------------------------------------------------------------------------
+!------------------------------------------------------------------------------
+! Modified by A. Porter, STFC.
```
Added.

TeranIvy commented on examples/dynamo/eg4/matrix_vector_kernel_mod.F90:
```diff
@@ -1,11 +1,43 @@
+! BSD 3-Clause License
 !
-!-------------------------------------------------------------------------------
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Added.

TeranIvy commented on examples/dynamo/eg4/mm_diagonal_kernel_mod.F90:
```diff
@@ -1,18 +1,48 @@
+! BSD 3-Clause License
 !
-!-------------------------------------------------------------------------------
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Added.

TeranIvy commented on examples/dynamo/eg4/solver_mod.x90:
```diff
@@ -1,8 +1,8 @@
 ! For further details please refer to the file LICENCE.original which you
 ! should have received as part of this distribution.
-!-------------------------------------------------------------------------------
+!-----------------------------------------------------------------------------
```
Added.

TeranIvy commented on examples/dynamo/eg7/alg.x90:
```diff
@@ -1,8 +1,8 @@
-!-------------------------------------------------------------------------------
+!-----------------------------------------------------------------------------
```
Added.

TeranIvy commented on examples/dynamo/eg7/columnwise_op_app_kernel_mod.F90:
```diff
@@ -1,14 +1,44 @@
-
-!> @brief Kernel which (incrementally) applies a columnwise assembled operator to a field
-
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Added.

TeranIvy commented on examples/dynamo/eg7/columnwise_op_asm_kernel_mod.F90:
```diff
@@ -1,14 +1,44 @@
-
-!> @brief Kernel which assembles a locally assembled matrix (LMA) into a columnwise assembled matrix (CMA)
-
+! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Added.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -540,6 +541,12 @@ is specified. Valid combinations are specified later in this section.
+* ``GH_READWRITE`` indicates that different iterations of a Kernel
+  update quantitites which do not share dofs, such as operators
+  and fields over relevant function spaces. Consequently, there is
+  no need for synchronisation (or colouring) when running these
```
Thanks. Elaborated on what happens for modification of purely discontinuous quantities and what happens additional gh_inc update.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -1811,6 +1816,13 @@ code that is added by PSyclone. See the ``README`` in the
 example.
 
+An example of applying boundary conditions to an operator is the kernel
+``enforce_operator_bc_kernel_mod.F90`` in
```
Changed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -1811,6 +1816,13 @@ code that is added by PSyclone. See the ``README`` in the
+``enforce_operator_bc_kernel_mod.F90`` in
+``<PSYCLONEHOME>/src/psyclone/tests/test_files/dynamo0p3`` directory.
+Since operators are discontinuous quantities, updating their values can be
+safely performed in parallel so ``GH_READWRITE`` access is used for this
```
Reworded.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -64,27 +64,26 @@ Before using the stub generator, PSyclone must be installed. If you
+PSyclone will be installed in a particular location on your machine,
+which will be referred to as ``<PSYCLONEINSTALL>`` directory. The stub
+generator script ``genkernelstub`` comes with the PSyclone installation.
+A quick check ``which genkernelstub`` should return the location of
```
Changed.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -64,27 +64,26 @@ Before using the stub generator, PSyclone must be installed. If you
-is installed (including the PSyclone directory itself) will be referred
-to as ``<PSYCLONEHOME>``.
+PSyclone will be installed in a particular location on your machine,
+which will be referred to as ``<PSYCLONEINSTALL>`` directory. The stub
```
Changed.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -304,28 +317,39 @@ they are invalid or because they contain functionality that is not
+    found 'any_space_1'"
+
+As noted above, if the Dynamo0.3 naming convention for module and type
+names is not followed, the stub generator will return with an error
```
Removed.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -540,6 +541,12 @@ is specified. Valid combinations are specified later in this section.
 
+* ``GH_READWRITE`` indicates that different iterations of a Kernel
+  update quantitites which do not share dofs, such as operators
+  and fields over relevant function spaces. Consequently, there is
```
Changed.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -64,27 +64,26 @@ Before using the stub generator, PSyclone must be installed. If you
+PSyclone will be installed in a particular location on your machine,
+which will be referred to as ``<PSYCLONEINSTALL>`` directory. The stub
+generator script ``genkernelstub`` comes with the PSyclone installation.
+A quick check ``which genkernelstub`` should return the location of
```
Changed.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -64,27 +64,26 @@ Before using the stub generator, PSyclone must be installed. If you
-is installed (including the PSyclone directory itself) will be referred
-to as ``<PSYCLONEHOME>``.
+PSyclone will be installed in a particular location on your machine,
+which will be referred to as ``<PSYCLONEINSTALL>`` directory. The stub
```
According to my installation and user experience, they are not the same directories. "PSYCLONEINSTALL" is where PSyclone scripts and libraries are located after installation (usually bin/ or Scripts/ directories for Python on Linux or Python Anaconda on Windows and /lib/python2.x/site-packages/ for both).  "PSYCLONEINSTALL" does not contain Fortran files used to test stub generation. 

"PSYCLONEHOME" is where PSyclone release is unpacked or where PSyclone is cloned and it contains Fortran files used to test stub generation. 

I can change this, however I do not think using "PSYCLONEHOME" for both locations is correct in general.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -51,12 +51,15 @@
 
-# constants
+# Function spaces (FS)
+# Continuous FS
```
Doh! Corrected.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -51,12 +51,15 @@
+# Continuous FS
 DISCONTINUOUS_FUNCTION_SPACES = ["w3", "wtheta", "w2v"]
-# space any_w2 can be w2, w2h or w2v
+# Discontinuous FS
```
Corrected.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -484,20 +492,26 @@ class DynArgDescriptor03(Descriptor):
     def __init__(self, arg_type):
+        '''
+        :param arg_type: dynamo0.3 argument type (scalar, field or operator)
+        :type arg_type: string
```
Updated.

TeranIvy commented on doc/dynamo0p3.rst:
```diff
@@ -555,10 +562,10 @@ For example:
-          Furthermore, this permission is currently restricted to integer
-          scalars (``GH_INTEGER``) as the LFRic infrastructure does not
+          Furthermore, this permission is currently restricted to real
+          scalars (``GH_REAL``) as the LFRic infrastructure does not
```
Thanks.

TeranIvy commented on doc/getting_going.rst:
```diff
@@ -286,6 +286,18 @@ If everything is working as expected then you should see output similar to:
 
    =================== 160 passed, 15 xfailed in 13.59 seconds ====================
 
+Most of the tests use Fortran source files in the
```
Thanks.

TeranIvy commented on doc/stub_gen.rst:
```diff
@@ -9,22 +9,22 @@ Quick Start
 1) Use an existing Kernel file or create a Kernel file containing a
```
Not at all :)

TeranIvy commented on doc/system_specific_setup.rst:
```diff
@@ -266,6 +266,11 @@ standards, use::
 
    > pep8 code.py
 
+.. warning::
```
I thought it would be good to have it.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -653,15 +675,33 @@ def __init__(self, arg_type):
                     format(VALID_FUNCTION_SPACE_NAMES, arg_type.args[2].name,
                            arg_type))
             self._function_space2 = arg_type.args[3].name
+            # Test allowed accesses for operators
```
Better to be safe than sorry :)

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -617,17 +624,32 @@ def __init__(self, arg_type):
                     "space ({0}) to have a 'gh_inc' access".
                     format(self._function_space1.lower()))
+            # TODO: extend for "gh_write"
+            if self._function_space1.lower() in CONTINUOUS_FUNCTION_SPACES \
```
Thanks.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4508,7 +4549,8 @@ def gen_code(self, parent):
         op_args = self.parent.args_filter(arg_types=VALID_OPERATOR_NAMES,
-                                          arg_accesses=["gh_read"])
+                                          arg_accesses=["gh_read",
+                                                        "gh_readwrite"])
```
Great.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4739,13 +4792,11 @@ def generate(self):
-            # TODO this access should really be "gh_readwrite". Support for
-            # this will be added under #25.
-            if op_arg.access != "gh_inc":
+            if op_arg.access != "gh_readwrite":
```
:)

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -566,37 +564,57 @@ def first_access(self, arg_name):
-        # as inc (shared update), write and read. A single argument may
-        # be accessed in different ways by different kernels.
+        # Get the lists of all kernel arguments that are accessed as
+        # inc (shared update), write, read and readwrite (not shared
```
Thank you for the suggestion. Implemented.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -1642,16 +1672,25 @@ def gen_code(self, parent):
-            # update scalar values appropriately
-            # HACK:TODO: update mapping to readwrite when it is supported
-            self._scalar.access = MAPPING_ACCESSES["inc"]
+            # Update scalar values appropriately
```
It took me some time and discussions with Andy to realise why this should be "readwrite" instead of "inc".

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -1719,9 +1758,10 @@ def __init__(self, field, check_dirty=True,
-            # update fields values appropriately
-            # HACK:TODO: update mapping to readwrite when it is supported
-            self._field.access = MAPPING_ACCESSES["inc"]
+            # Update fields values appropriately
```
Same as above.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -1,11 +1,7 @@
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
-# Copyright (c) 2017, Science and Technology Facilities Council
```
Not at all :)

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -566,37 +564,57 @@ def first_access(self, arg_name):
-        # do not appear in the list of those that are written.
+        # sum_args and readwrite_args behave as if they are write_args from
+        # the PSy-layer's perspective.
+        write_args += sum_args + readwrite_args
```
Thank you for the suggestion. I had the feeling I made a lot of redundant changes but I went for what I thought was an extra safe approach. I implemented your suggestions and updated the comments to reflect that readwrite_args are treated together with inc_args.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -2425,7 +2465,18 @@ def incremented_arg(self, mapping={}):
+        '''
         assert mapping != {}, "psyGen:Kern:written_arg: Error - a "\
             "mapping must be provided"
         for access in ["write", "readwrite"]:
```
There is a function above which returns incremented_arg. Are you suggesting that all modified args could be grouped together?

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -2550,9 +2613,11 @@ def __init__(self, call, arg_info, access):
         # MAPPING_ACCESSES specified in the dynamo0p3 file which
         # overide the default ones in this file.
         self._write_access_types = [MAPPING_ACCESSES["write"],
+                                    MAPPING_ACCESSES["readwrite"],
```
Good.

TeranIvy commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -650,6 +662,11 @@ def test_cma_asm():
                 "cbanded_map_any_space_1_lma_op1, ndf_any_space_2_lma_op1, "
                 "cbanded_map_any_space_2_lma_op1)") in code
 
+        if utils.TEST_COMPILE:
```
Thanks.

TeranIvy commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4878,20 +4967,37 @@ def test_loop_fusion_different_loop_depth():
     rc_trans.apply(schedule.children[0], depth=3)
     f_trans = DynamoLoopFuseTrans()
     with pytest.raises(TransformationError) as excinfo:
+        # f1 and f2 have read accesses (readwrite and read) so there
```
What I was trying to note here is that the loops will move in the schedule due to additional halo exchange in f1 RW field. The test is still for different loop bounds names. As the comment was misleading, I removed the reference to dependencies and halo exchanges and just left that 2 and 3 are indices of loops in the schedule.

TeranIvy commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4799,6 +4854,40 @@ def test_rc_vector_reader_halo_remove():
+    assert result.count("is_dirty") == 9
+    assert result.count("halo_exchange") == 9
+
+
```
Increased the depth of the reader loop to 2 and added checks that the additional f1 halo exchanges do not have if tests around them and that they are of depth 2.

TeranIvy commented:
Hi @rupertford! I hope I addressed all remarks related to the existing code. I still need to add the new requested tests. This was delayed due to emergency Python installation for LFRic (long story). Hopefully I will complete the additional tests tomorrow.

TeranIvy commented:
**Update**

a) "Test for disallowing readwrite with colouring" 
I extended `test_colour_discontinuous()` in `dynamo0p3_transformations_test.py` to test a discontinuous readwriter as well.

b) "Test for readwrite followed by a read gives an inout intent and a read followed by a readwrite gives an out intent."
I am afraid I do not understand the request, as `def gen_code(self, parent)` in `class DynInvoke(Invoke)` replaces `out` intent for fields and operators with `inout` so there are no examples of generated code with `intent(out)` for fields. @rupertford, could you please elaborate on what you had in mind?

c) "That readwrite works OK with OMP directives (just one example would be fine)."
Added two tests with w3 discontinuous readwriter: `test_loop_fuse_omp_rwdisc` and `test_omp_region_omp_do_rwdisc`. There is one new file which is checked for compilation.

d) `test_written_arg` in `psyGen_test.py` had incorrect mapping for readwrite so I corrected it. I also added `test_incremented_arg`.

Brought the branch up to date. Ready for the second review.

TeranIvy commented:
**Note for the next round of review:** I just found out that there is a check for discontinuous field which only uses `w3` in the `class DynamoOMPParallelLoopTrans` (file `transformations.py`). The relevant lines need to be corrected from 

```
        if node.field_space.orig_name != "w3":
            if node.loop_type is not 'colour' and node.has_inc_arg():
```
to

```
        if node.field_space.orig_name not in DISCONTINUOUS_FUNCTION_SPACES:
            if node.loop_type is not 'colour' and node.has_inc_arg():
```
